src/coreComponents/common/TypeDispatch.hpp

```diff
@@ -0,0 +1,341 @@
+template< int M, int N >
+using DimsRange = camp::as_list< internal::Increment< camp::make_idx_seq_t< N - M + 1 >, M > >;
+
+/**
+ * @brief Generate a list of types representing array dimensionality exactly @p N.
+ */
+template< int N >
+using DimsSingle = DimsRange< N, N >;
+
+/**
+ * @brief Generate a list of types representing array dimensionalities up to (and including) @p N.
+ */
+template< int N >
+using DimsUpTo = DimsRange< 1, N >;
+
+/**
+ * @brief List of real-valued array types typically used in GEOSX (dimensions up to 4).
+ */
+using RealArrays = ArrayTypes< RealTypes, DimsUpTo< 4 > >;
+
+/**
+ * @brief List of integer-valued array types typically used in GEOSX (dimensions 1 and 2).
+ */
+using IntegralArrays = ArrayTypes< IntegralTypes, DimsUpTo< 2 > >;
+
+/**
+ * @brief List of all array types (real- and integer-valued) typically used in GEOSX.
+ */
+using StandardArrays = Join< RealArrays, IntegralArrays >;
+
+namespace internal
+{
+
+template< typename LAMBDA >
+bool dispatchRecursive( TypeList<>,
+                        std::type_index const type,
+                        LAMBDA && lambda )
+{
+  GEOSX_UNUSED_VAR( type, lambda )
+  return false;
+}
+
+template< typename T, typename ... REST, typename LAMBDA >
+bool dispatchRecursive( TypeList< T, REST... >,
+                        std::type_index const type,
+                        LAMBDA && lambda )
+{
+  if( type == typeid( T ) )
+  {
+    lambda( T{} );
+    return true;
+  }
+  return dispatchRecursive( TypeList< REST... >{}, type, std::forward< LAMBDA >( lambda ) );
+}
+
+template< typename ... Ts, std::size_t ... Is >
+std::unordered_map< std::type_index, std::size_t > const &
+getTypeIndexMap( TypeList< Ts... >,
+                 std::index_sequence< Is... > )
+{
+  static std::unordered_map< std::type_index, std::size_t > const result{ { typeid( Ts ), Is } ... };
+  return result;
+}
+
+template< typename ... Ts, typename LAMBDA >
+bool dispatchViaTable( TypeList< Ts... > const types,
+                       std::type_index const type,
+                       LAMBDA && lambda )
+{
+  // Initialize a table of handlers, once per unique combination of type list and lambda
+  using Handler = void (*)( LAMBDA && );
+  static Handler const handlers[] = { []( LAMBDA && f ){ f( Ts{} ); } ... };
+
+  // Initialize a hashmap of std::type_index to contiguous indices, once per unique type list
+  auto const & indexMap = getTypeIndexMap( types, std::index_sequence_for< Ts... >{} );
+  auto const it = indexMap.find( type );
+  if( it != indexMap.end() )
+  {
+    GEOSX_ASSERT_GT( sizeof...( Ts ), it->second ); // sanity check
+    handlers[ it->second ]( std::forward< LAMBDA >( lambda ) );
+    return true;
+  }
+  return false;
+}
+
+template< typename LAMBDA >
+bool dispatchViaTable( TypeList<>,
+                       std::type_index const type,
+                       LAMBDA && lambda )
+{
+  // Needed to avoid constructing an zero-size array
+  GEOSX_UNUSED_VAR( type, lambda )
```

Comment:
Do we actually use this somewhere?

Reply:
Only in unit tests :)

Reply:
If we don't see a need for it in the main code I'd replace the body of the function with a static_assert, just so that you get a decent compilation error instead of something gross.